### PR TITLE
[ScatterND]: code generation & improved verification code

### PR DESCRIPTION
**Pull request by etiotto**
Implement support for the ONNX `ScatterND` operator:

- [x] verification code (diagnose operator constraints)
- [x] shape inference (trivial)
- [x] codegen support
- [x] added lit tests to verify constraint diagnostics 
- [x] added lit test to verify code generation
- [x] enable end-to-end test (backend test)

Signed-off-by: Ettore Tiotto <etiotto@ca.ibm.com>

### CODE REVIEW

File: src/Conversion/ONNXToKrnl/Tensor/ScatterND.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    Value indices = operandAdaptor.indices();
+    auto dataType = data.getType().cast<ShapedType>();
+    auto indicesType = indices.getType().cast<ShapedType>();
+    auto updatesType = updates.getType().cast<ShapedType>();
+    int64_t dataRank = dataType.getRank();
+    int64_t updatesRank = updatesType.getRank();
+    int64_t indicesRank = indicesType.getRank();
+
+    assert(dataRank >= 1 && "The rank of 'data' must be >= 1");
+    assert(indicesRank >= 1 && "The rank of 'indices' must be >= 1");
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Tensor/ScatterND.cpp:44]: I assume that these input dependent assert where tested in the verifier and are here only for redundancy (which is perfectly fine IMO).
Code below seems to confirm that the verifiers indeed tested these conditions.

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -4462,55 +4461,120 @@ LogicalResult ONNXScatterElementsOp::verify() {
...
+          std::to_string(dataShape[i]));
+  }
 
   return success();
 }
 
 LogicalResult ONNXScatterNDOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  if (!data().getType().isa<mlir::RankedTensorType>())
+  if (!data().getType().isa<mlir::ShapedType>())
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNXOps.cpp:4558]: thoughts we converged to use "has shape and rank"

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM

File: src/Conversion/ONNXToKrnl/Tensor/ScatterND.cpp (hunk 1 above)
Comment (etiotto) [reply src/Conversion/ONNXToKrnl/Tensor/ScatterND.cpp:44]: Yes that is correct.  The idea is to diagnose error conditions in the verifier, and in shape inference defensively assert that the "code" is as expected.

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 2 above)
Comment (etiotto) [reply src/Dialect/ONNX/ONNXOps.cpp:4558]: Good point. Going to fix this.

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 3]
```diff
@@ -107,19 +107,18 @@ static LogicalResult shapeHelperInferMultipleShapes(
 // Handle shape inference for numpy style broadcasting operators.
 template <class OP, class ADAPTOR>
 static LogicalResult inferShapeForBroadcastingOps(
-    OP *op, Type elementType = nullptr) {
-  assert(op && "Expecting a non-null pointer");
-  ADAPTOR operandAdaptor(*op);
+    OP &op, Type elementType = nullptr) {
```
Comment (etiotto) [inline src/Dialect/ONNX/ONNXOps.cpp:110]: Note: passing `op` as a reference allows us to avoid assertion check that op is a valid pointer.

**Pull request merged**